### PR TITLE
accountsservice: refactor package and service

### PR DESCRIPTION
**Pull request by groxxda**
###### Motivation for this change

This makes the following changes to accountsservice:
- Add `nixbld*` to the blacklist (so they will not be listed on login screen, etc)
- Fix a few path to `/run/current-system/sw/...`
- Remove the functions that manipulate `/etc/{passwd, ...}` when `mutableUsers` is set to `false`

I left the functions that modify `/var/lib/AccountsService/` active because these files are not managed by nix and are relevant for desktop environments (icon, etc.)
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

before (`mutableUsers = *`):

```
$ dbus-send --print-reply --system --dest=org.freedesktop.Accounts /org/freedesktop/Accounts org.freedesktop.Accounts.CreateUser string:abc string:"max mustermann" int32:1
Error org.freedesktop.Accounts.Error.Failed: running '/usr/sbin/useradd' failed: Failed to execute child process "/usr/sbin/useradd" (No such file or directory)
```

now (`mutableUsers = true`):

```
$ dbus-send --print-reply --system --dest=org.freedesktop.Accounts /org/freedesktop/Accounts org.freedesktop.Accounts.CreateUser string:abc string:"max mustermann" int32:1
method return time=1464780613.660190 sender=:1.267 -> destination=:1.269 serial=15 reply_serial=2
   object path "/org/freedesktop/Accounts/User1001"
% grep abc /etc/passwd
abc:x:1001:100:max mustermann:/home/abc:/run/current-system/sw/bin/zsh
```

now (`mutableUsers = false`):

```
$ sudo dbus-send --print-reply --system --dest=org.freedesktop.Accounts /org/freedesktop/Accounts org.freedesktop.Accounts.CreateUser string:abc string:"max mustermann" int32:1
Error org.freedesktop.Accounts.Error.NotSupported: Modifying users not supported without users.mutableUsers
```

CC: @pSub

### CODE REVIEW

**Comment by groxxda** [comment]
Not sure why travis complains `no configure script, doing nothing` :confused:

**Comment by dezgeg** [comment]
I have a feeling some of the latest changes to the Travis-CI scripts broke something.

cc @domenkozar

**Comment by domenkozar** [comment]
Not really, linux part was untouched.

**Comment by joachifm** [comment]
This makes sense for NixOS and looks like a clear improvement, but it always makes me a little sad to see a package effectively become NixOS only.

**Comment by lethalman** [comment]
What about using an environment variable instead of relying on a compile-time patch?

**Comment by groxxda** [comment]
@joachifm would replacing with `${util-linux}/bin/nologin` be an option? I guess not since util-linux itself seems nixos-only :-/
@lethalman can you elaborate how that would work?

**Comment by lethalman** [comment]
@groxxda if accountsservice is started with NIXOS_USERS_PURE=1 then it will throw the exceptions, otherwise not. That is, at runtime, instead of compile time.

**Comment by groxxda** [comment]
@lethalman: that way we would can share the same binary package for both use cases, in exchange of having to maintain a more complex patch, but is there something else we gain?

**Comment by lethalman** [comment]
@groxxda no I think that's all, note that the more complex patch would be just an `if (getenv("NIXOS_USERS_PURE")) throw`.

**Comment by joachifm** [comment]
@groxxda any progress on making this more compatible with !NixOS or is it too gnarly to get working cleanly? In the latter case, I guess there's no other way ...

**Comment by groxxda** [comment]
I just implemented the environment variable approach and also bumped the accountsservice to latest version.

* Tested locally
* When changing `mutableUsers` no rebuild should occur anymore

Should be good to merge now.

CC @joachifm @lethalman @fpletz

File: nixos/modules/services/desktops/accountsservice.nix [hunk 1]
```diff
@@ -35,6 +35,14 @@ with lib;
     services.dbus.packages = [ pkgs.accountsservice ];
 
     systemd.packages = [ pkgs.accountsservice ];
+
+    systemd.services.accounts-daemon= {
+
+      wantedBy = [ "graphical.target" ];
```
Comment (lethalman) [inline nixos/modules/services/desktops/accountsservice.nix]: Why is this necessary? The service should be auto activated by dbus.

File: pkgs/development/libraries/accountsservice/default.nix [hunk 2]
```diff
@@ -3,22 +3,29 @@
...
+    sha256 = "0zh0kjpdc631qh36plcgpwvnmh9wj8l5cki3aw5r09w6y7198r75";
   };
 
   buildInputs = [ pkgconfig glib intltool libtool makeWrapper
                   gobjectIntrospection polkit systemd ];
 
   configureFlags = [ "--with-systemdsystemunitdir=$(out)/etc/systemd/system"
                      "--localstatedir=/var" ];
+  prePatch = ''
+    substituteInPlace src/daemon.c --replace '"/usr/sbin/' '"/run/current-system/sw/sbin/'
```
Comment (lethalman) [inline pkgs/development/libraries/accountsservice/default.nix]: Maybe instead of relying on /run/current-system you can point to the exact binaries instead. Like ${coreutils}/bin/cat .

File: nixos/modules/services/desktops/accountsservice.nix (hunk 1 above)
Comment (groxxda) [inline nixos/modules/services/desktops/accountsservice.nix]: From the original service file:
```
[Install]
# We pull this in by graphical.target instead of waiting for the bus
# activation, to speed things up a little: gdm uses this anyway so it is nice
# if it is already around when gdm wants to use it and doesn't have to wait for
# it.
WantedBy=graphical.target
```

**Comment by groxxda** [comment]
@lethalman : changed to exact store paths.
I have no strong opinion on the `wantedBy`, but included it because upstream does so too

**Comment by lethalman** [comment]
Looks good thanks. I hardly believe those shadow programs work without the suid bit, but it's better than before :)

**Comment by domenkozar** [comment]
Shall we cherry-pick to 16.09?

**Comment by lethalman** [comment]
@domenkozar I haven't tested it and it's not top priority, so I'd say no.

**Comment by fpletz** [comment]
I think it's worth backporting to 16.09 because it enhances the default user experience of new users by removing the nixbld users from the login screen. I'm gonna test this and would like to backport.
